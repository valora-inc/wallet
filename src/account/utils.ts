import { parsePhoneNumber } from '@celo/phone-utils'

const ADDRESS_LENGTH = 42 // TODO(ACT-1173): see if this can be replaced with a viem helper

export const isAddressFormat = (content: string): boolean => {
  return content.startsWith('0x') && content.length === ADDRESS_LENGTH
}

export function getPhoneNumberDetails(
  phoneNumber: string,
  countryCallingCode: string,
  countryCodeAlpha2: string
) {
  const phoneDetails = parsePhoneNumber(phoneNumber, countryCallingCode)

  if (phoneDetails) {
    return {
      // Show international display number to avoid confusion
      internationalPhoneNumber: phoneDetails.displayNumberInternational,
      e164Number: phoneDetails.e164Number,
      isValidNumber: true,
      countryCodeAlpha2: phoneDetails.regionCode!,
    }
  } else {
    return {
      internationalPhoneNumber: phoneNumber,
      e164Number: '',
      isValidNumber: false,
      countryCodeAlpha2,
    }
  }
}
