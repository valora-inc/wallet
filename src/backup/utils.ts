import {
  CELO_DERIVATION_PATH_BASE,
  generateKeys,
  MnemonicLanguages,
} from '@celo/cryptographic-utils'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import * as bip39 from 'react-native-bip39'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { useDispatch, useSelector } from 'src/redux/hooks'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { ETHEREUM_DERIVATION_PATH } from 'src/web3/consts'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

const MNEMONIC_STORAGE_KEY = 'mnemonic'

export async function generateKeysFromMnemonic(mnemonic: string) {
  const wordCount = countMnemonicWords(mnemonic)
  const derivationPath = wordCount === 24 ? CELO_DERIVATION_PATH_BASE : ETHEREUM_DERIVATION_PATH
  return generateKeys(mnemonic, undefined, undefined, undefined, bip39, derivationPath)
}

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Recovery Phrase',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function isValidBackupPhrase(phrase: string) {
  const allowedPhraseLengths = [12, 24]
  const phraseLength = countMnemonicWords(formatBackupPhraseOnEdit(phrase))
  return !!phrase && allowedPhraseLengths.includes(phraseLength)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
