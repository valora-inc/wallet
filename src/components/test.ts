function groupNumber(value: string) {
  return value.replace(/\B(?=(\d{3})+(?!\d))(?<!\.\d*)/g, 'group')
}

function actualTest(value: string, decimalSeparator: string, groupingSeparator: string) {
  return groupNumber(value).replaceAll('.', decimalSeparator).replaceAll('group', groupingSeparator)
}

console.log(actualTest('1.234.567,9999', ',', '.')) // 1,234,567.9999
