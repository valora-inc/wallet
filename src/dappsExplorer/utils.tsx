import { DappV2WithCategoryNames } from 'src/dapps/types'

/***
 * Calculates a score based on the search term and dapp
 * - +2 if the name matches the search term in a case sensitive way
 * - +1 if the name matches the search term in a case insensitive way
 * - +1 if the description matches the search term in a case insensitive way
 * - +1 if the category matches the search term in a case insensitive way
 * @param dapp: dapp with categoryNames
 * @param searchTerm: the user input search string
 * @returns {number} the score of the dapp 0 - 5
 */
export const calculateSearchScore = (dapp: DappV2WithCategoryNames, searchTerm: string) => {
  if (searchTerm === '') return 0
  const nameMatchingCaseScore = dapp.name.includes(searchTerm) ? 2 : 0
  const nameScore = dapp.name.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const descriptionScore = dapp.description.toLowerCase().includes(searchTerm.toLowerCase()) ? 1 : 0
  const categoryScore = dapp.categoryNames.some((category) =>
    category.toLowerCase().includes(searchTerm.toLowerCase())
  )
    ? 1
    : 0
  return nameMatchingCaseScore + nameScore + descriptionScore + categoryScore
}
