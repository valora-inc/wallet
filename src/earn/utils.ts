import BigNumber from 'bignumber.js'
import { getFeatureGate } from 'src/statsig'
import { StatsigFeatureGates } from 'src/statsig/types'
import { SwapTransaction } from 'src/swap/types'
import { Network, NetworkId } from 'src/transactions/types'
import { INTERNAL_RPC_SUPPORTED_NETWORKS } from 'src/viem'
import { networkIdToNetwork } from 'src/web3/networkConfig'

/**
 * Helper function to check if gas is subsidized for the given networkId
 * @param networkId - networkId to check
 * @returns boolean
 */
export function isGasSubsidizedForNetwork(networkId: NetworkId) {
  return (
    (INTERNAL_RPC_SUPPORTED_NETWORKS as readonly Network[]).includes(
      networkIdToNetwork[networkId]
    ) && getFeatureGate(StatsigFeatureGates.SUBSIDIZE_STABLECOIN_EARN_GAS_FEES)
  )
}

/**
 * Gets the swap to amount in decimals given a swap transaction and the swap
 * from amount in decimals. This could also be taken from the buy amount, but
 * that is in wei and we would need the token decimals. This follows what is
 * done in the SwapScreen
 * @param object.swapTransaction - swap transaction object
 * @param object.fromAmount - swap from amount in decimals (usually from the user input)
 * @returns swap to amount in decimals
 */
export function getSwapToAmountInDecimals({
  swapTransaction,
  fromAmount,
}: {
  swapTransaction: SwapTransaction
  fromAmount: BigNumber
}) {
  return fromAmount.multipliedBy(swapTransaction.price)
}
