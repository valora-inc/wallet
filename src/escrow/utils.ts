import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToE164NumberSelector } from 'src/identity/selectors'
import { ContactRecipient } from 'src/recipients/recipient'
import { useSelector } from 'src/redux/hooks'

export const useEscrowPaymentRecipient = (payment: EscrowedPayment): ContactRecipient => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToE164Number = useSelector(identifierToE164NumberSelector)
  const recipientCache = useSelector((state) => state.recipients.phoneRecipientCache)
  const phoneNumber = identifierToE164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber] ?? {
    name: phoneNumber,
    e164PhoneNumber: phoneNumber,
    contactId: '',
  }

  return recipient
}
