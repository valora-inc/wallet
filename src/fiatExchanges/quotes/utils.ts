import { SettlementEstimation, SettlementTime } from 'src/fiatExchanges/quotes/constants'
import Logger from 'src/utils/Logger'

const TAG = 'fiatExchanges/quotes/utils'

export const getSettlementTimeString = (
  settlementEstimation: SettlementEstimation,
  settlementTimeStrings: Record<SettlementTime, string>
) => {
  switch (settlementEstimation.settlementTime) {
    case SettlementTime.LESS_THAN_ONE_HOUR:
      return { timeString: settlementTimeStrings[SettlementTime.LESS_THAN_ONE_HOUR] }
    case SettlementTime.LESS_THAN_X_HOURS:
    case SettlementTime.LESS_THAN_X_DAYS:
      return {
        timeString: settlementTimeStrings[settlementEstimation.settlementTime],
        upperBound: settlementEstimation.upperBound,
      }
    case SettlementTime.X_TO_Y_HOURS:
    case SettlementTime.X_TO_Y_DAYS:
      return {
        timeString: settlementTimeStrings[settlementEstimation.settlementTime],
        lowerBound: settlementEstimation.lowerBound,
        upperBound: settlementEstimation.upperBound,
      }
    default:
      // Should not be reachable
      Logger.warn(TAG, 'Unexpected settlement time value in settlement estimation')
      return {}
  }
}
