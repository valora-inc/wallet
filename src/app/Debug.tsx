import Clipboard from '@react-native-clipboard/clipboard'
import * as React from 'react'
import { StyleSheet, Text } from 'react-native'
import DeviceInfo from 'react-native-device-info'
import { SafeAreaView } from 'react-native-safe-area-context'
import { connect } from 'react-redux'
import Button, { BtnTypes } from 'src/components/Button'
import { noHeader } from 'src/navigator/Headers'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { RootState } from 'src/redux/reducers'
import Logger from 'src/utils/Logger'
import { currentAccountSelector } from 'src/web3/selectors'
import { getLatestBlock } from 'src/web3/utils'
import { Statsig } from 'statsig-react-native'

interface State {
  latestBlockNumber: number
}

export class Debug extends React.Component<RootState, State> {
  static navigationOptions = noHeader

  state = {
    latestBlockNumber: 0,
  }

  async componentDidMount() {
    await this.updateLatestBlock()
  }

  updateLatestBlock = async () => {
    const latestBlock = await getLatestBlock()
    this.setState({
      latestBlockNumber: (latestBlock && latestBlock.number) || 0,
    })
  }

  onClickText = (...text: Array<string | null>) => {
    return () => {
      Logger.showMessage('Copied to Clipboard')
      Clipboard.setString(text.join(', '))
    }
  }

  onClickEmailLogs = async () => {
    navigate(Screens.SupportContact)
  }

  render() {
    const { latestBlockNumber } = this.state
    const pincodeType = this.props.account.pincodeType
    const address = currentAccountSelector(this.props)
    const phoneNumber = this.props.account.e164PhoneNumber
    const version = DeviceInfo.getVersion()
    const buildNumber = DeviceInfo.getBuildNumber()
    const apiLevel = DeviceInfo.getApiLevelSync()
    const deviceId = DeviceInfo.getDeviceId()
    const stableId = Statsig.getStableID()

    return (
      <SafeAreaView style={styles.container}>
        <Text
          onPress={this.onClickText(deviceId, phoneNumber)}
          style={styles.singleLine}
        >{`Device Id: ${deviceId} | Phone Number: ${phoneNumber}`}</Text>
        <Text
          onPress={this.onClickText(version, buildNumber, String(apiLevel))}
          style={styles.singleLine}
        >{`Version: ${version} | Build Number: ${buildNumber} | Api Level: ${apiLevel}`}</Text>
        <Text style={styles.singleLine}>{`Pin Type: ${pincodeType}`}</Text>
        <Text
          onPress={this.onClickText(address)}
          style={styles.singleLine}
        >{`Address: ${address}`}</Text>
        <Text onPress={this.onClickText(stableId)} style={styles.singleLine}>
          {`Statsig Stable ID: ${stableId}`}
        </Text>
        <Text style={styles.singleLine}>{`Latest Block: ${latestBlockNumber}`}</Text>
        <Button
          onPress={this.onClickEmailLogs}
          text={'Email logs to support'}
          type={BtnTypes.PRIMARY}
          style={styles.button}
        />
      </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingVertical: 5,
    paddingHorizontal: 10,
    justifyContent: 'space-between',
  },
  singleLine: {
    marginTop: 5,
    fontSize: 11,
  },
  button: {
    marginHorizontal: 0,
    marginTop: 10,
    marginBottom: 0,
  },
})

export default connect<RootState, {}, {}, RootState>((state: RootState) => state)(Debug)
