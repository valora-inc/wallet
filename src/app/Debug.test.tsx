import { render } from '@testing-library/react-native'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import Debug from 'src/app/Debug'
import { createMockStore } from 'test/utils'

jest.mock('statsig-react-native', () => ({
  Statsig: {
    getStableID: () => 'stableId',
  },
}))

describe('Debug', () => {
  it('renders correctly', () => {
    const tree = render(
      <Provider store={createMockStore()}>
        <Debug />
      </Provider>
    )
    expect(tree).toMatchSnapshot()
  })
})
