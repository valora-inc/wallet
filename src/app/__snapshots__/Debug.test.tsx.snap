// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Debug renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "flex": 1,
      "justifyContent": "space-between",
      "paddingHorizontal": 10,
      "paddingVertical": 5,
    }
  }
>
  <Text
    onPress={[Function]}
    style={
      {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Device Id: someDeviceId | Phone Number: +14155556666
  </Text>
  <Text
    onPress={[Function]}
    style={
      {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Version: 0.0.1 | Build Number: 1 | Api Level: -1
  </Text>
  <Text
    style={
      {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Pin Type: Unset
  </Text>
  <Text
    onPress={[Function]}
    style={
      {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Address: 0x0000000000000000000000000000000000007e57
  </Text>
  <Text
    style={
      {
        "fontSize": 11,
        "marginTop": 5,
      }
    }
  >
    Latest Block: 0
  </Text>
  <View
    style={
      [
        {
          "flex": 1,
        },
        {
          "flex": 1,
          "marginTop": 8,
        },
      ]
    }
  >
    <Text>
      React-Native Logs
    </Text>
    <RCTScrollView
      onContentSizeChange={[Function]}
    >
      <View>
        <Text
          onPress={[Function]}
          style={
            {
              "fontSize": 10,
            }
          }
        />
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      [
        {
          "flexDirection": "row",
        },
        {
          "marginBottom": 0,
          "marginHorizontal": 0,
          "marginTop": 10,
        },
      ]
    }
  >
    <View
      style={
        [
          {
            "overflow": "hidden",
          },
          {
            "borderRadius": 100,
          },
        ]
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          {
            "attribute": "selectableItemBackground",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            {
              "alignItems": "center",
              "backgroundColor": "#1AB775",
              "flexDirection": "row",
              "height": 48,
              "justifyContent": "center",
              "minWidth": 120,
              "opacity": undefined,
              "paddingHorizontal": 24,
              "paddingVertical": 5,
            },
            undefined,
          ]
        }
      >
        <Text
          maxFontSizeMultiplier={1}
          style={
            {
              "color": "#FFFFFF",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 16,
              "lineHeight": 22,
              "marginLeft": 0,
              "marginRight": 0,
            }
          }
        >
          Email logs to support
        </Text>
      </View>
    </View>
  </View>
</RNCSafeAreaView>
`;
